pyro/ops/welford.py

```diff
@@ -0,0 +1,46 @@
+import torch
+
+
+class WelfordCovariance(object):
+    """
+    Implements Welford's online scheme for estimating (co)variance (see :math:`[1]`).
+    Useful for adapting diagonal and dense mass structures for HMC.
+
+    **References**
+
+    [1] `The Art of Computer Programming`,
+    Donald E. Knuth
+    """
+    def __init__(self, diagonal=True):
+        self.diagonal = diagonal
+        self.reset()
+
+    def reset(self):
+        self._mean = 0.
+        self._m2 = 0.
+        self.n_samples = 0
+
+    def update(self, sample):
+        self.n_samples += 1
+        delta_pre = sample - self._mean
+        self._mean = self._mean + delta_pre / self.n_samples
+        delta_post = sample - self._mean
+
+        if self.diagonal:
+            self._m2 += delta_pre * delta_post
+        else:
+            self._m2 += delta_pre * delta_post.reshape(-1, 1)
```

Comment:
Use `torch.matmul` is faster (I made a benchmark [here](https://gist.github.com/fehiepsi/6c5c769af2b2ca9e29c301ba30b98f6d)). In addition, to add an additional dimension, I think that using `.unsqueeze()` is better (faster) than `.reshape()`.